Feature - add message context filter to the message handler registration

stijnmoreels opened the pull request:
Relates to #26

stijnmoreels commented:
Seems like the build is failing on master too, so could be that it's not related with these changes here.

tomkerkhove commented on docs/preview/features/message-pumps/customization.md:
````diff
@@ -40,4 +40,34 @@ public class OrderMessagePump : MessagePump
+}
+```
+
+> Note that the order in which the message handlers are registered is importent in the message processing.
````
```suggestion
> Note that the order in which the message handlers are registered is important in the message processing.
```

tomkerkhove commented on docs/preview/features/message-pumps/customization.md:
```diff
@@ -40,4 +40,34 @@ public class OrderMessagePump : MessagePump
 
+## Control message context filtering
+
+When registering a new `IMessageHandler<>`, one can choose to add an additional filter on the message context used in that handler.
```
What about this?
```suggestion
When registering a new message handler, one can opt-in to add a filter on the message context which filters out messages that are not needed to be processed.

This can be useful when you are sending different message types on the same queue. Another use-case is being able to handle different versions of the same message type which have different contracts because you are migrating your application.
```

tomkerkhove commented on docs/preview/features/message-pumps/customization.md:
````diff
@@ -40,4 +40,34 @@ public class OrderMessagePump : MessagePump
 }
 ```
 
+## Control message context filtering
````
What do you think of one of these?
```suggestion
## Filter messages based on message context
```

```suggestion
## Process subset of messages with message context filtering
```

tomkerkhove commented on docs/preview/features/message-pumps/customization.md:
````diff
@@ -40,4 +40,34 @@ public class OrderMessagePump : MessagePump
+```csharp
+public class OrderMessageHandler : IMessageHandler<Order>
+{
+    public async Task ProcessMessageAsync(Order order, MessageContext context, ...)
````
It's been a while so here is a stupid question - They were able to use different MessageContext type as well, right? Like the Service bus one?

stijnmoreels replied:
Not a stupid question 😉 , yest that's true, one can register message handlers with custom and built-in message context, altough I didn't implement or test this just yet in this PR. I thought it was a good start to first begin with the 'baseline' for this message context filtering.

Also why this PR is not closing the related issue 🤓 .

tomkerkhove replied:
Ok, thanks!

stijnmoreels replied:
But as explained below, I added some extensions for this too bc I needed it for the integration tests.

stijnmoreels commented:
> This looks good to me, added a few comments.
> 
> Can we introduce some integration tests for this please?

Yes, of course. What would you have in mind?
* Registering two handlers, only one with the correct context type
* Registering multiple handlers with the same context type, only one with the correct context predicate

stijnmoreels commented:
> > This looks good to me, added a few comments.
> > Can we introduce some integration tests for this please?
> 
> Yes, of course. What would you have in mind?
> 
> * Registering two handlers, only one with the correct context type
> * Registering multiple handlers with the same context type, only one with the correct context predicate

Ok, I added these two tests + required me to add the extensions on the service bus message handlers, so that's added no too.

tomkerkhove commented:
> > This looks good to me, added a few comments.
> > Can we introduce some integration tests for this please?
> 
> Yes, of course. What would you have in mind?
> 
> * Registering two handlers, only one with the correct context type
> * Registering multiple handlers with the same context type, only one with the correct context predicate

Just some tests to verify that the correct handler is processing messages if the criteria is met

stijnmoreels commented:
> > > This looks good to me, added a few comments.
> > > Can we introduce some integration tests for this please?
> > 
> > 
> > Yes, of course. What would you have in mind?
> > 
> > * Registering two handlers, only one with the correct context type
> > * Registering multiple handlers with the same context type, only one with the correct context predicate
> 
> Just some tests to verify that the correct handler is processing messages if the criteria is met

Yes, added two tests to test selection based on type and predicate.

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -728,5 +728,49 @@ public static class IServiceCollectionExtensions
+        /// <typeparam name="TMessage">The type of the message that the message handler will process.</typeparam>
+        /// <param name="services">The collection of services to use in the application.</param>
+        /// <param name="messageContextFilter">The function that determines if the message handler should handle the message based on the context.</param>
+        public static IServiceCollection WithServiceBusMessageHandler<TMessageHandler, TMessage>(
```
We should decouple this from Service Bus and only give them `MessageContext`

tomkerkhove commented on src/Arcus.Messaging.Pumps.ServiceBus/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -728,5 +728,49 @@ public static class IServiceCollectionExtensions
+        /// <param name="services">The collection of services to use in the application.</param>
+        /// <param name="messageContextFilter">The function that determines if the message handler should handle the message based on the context.</param>
+        /// <param name="implementationFactory">The function that creates the message handler.</param>
+        public static IServiceCollection WithServiceBusMessageHandler<TMessageHandler, TMessage>(
```
We should decouple this from Service Bus and only give them `MessageContext`

stijnmoreels replied:
Why is that? Isn't that the point of registering a service bus message handler? If people want a different context, shouldn't they use the `.WithMessageHandler` instead?

tomkerkhove replied:
This is what we want to avoid:
![image](https://user-images.githubusercontent.com/4345663/86323929-05d41200-bc3e-11ea-9267-ad6521586621.png)

Meaning, people need to have the opportunity to be able to use message handler with filtering that does not tightly-couple it to Service Bus. Do that mean we should remove the current on? No, but we need to have the filtering capability on `WithMessageHandler` at least.

That's why we decoupled it a while back, to make sure the same message handler can be used with another queuing tech.

stijnmoreels replied:
Yes, and the context filtering is available on the `.WithMessageHandler`, the `.WithServiceBusMessageHandler` only adds the specific service bus message context to the handler so we don't have to specify:

```csharp
.WithMessageHandler<OrdersMessageHandler, Order, ServiceBusMessageContext>()
```

stijnmoreels replied:
So, I'm not sure what I should remove/change here and if I implement something wrong here maybe?

tomkerkhove replied:
I've misread, didn't notice the other extensions sorry! No action to be taken then but I would at least add some tests using the non-SB approach since we're replacing those

stijnmoreels replied:
Yes, updated the existing test to verify more thoroughly the predication selection.

stijnmoreels commented:
> I might be taking this a bit too far but I'm missing an integration test where we:
> 
> 1. Send a message with a given property in context
> 2. Have multiple handlers configured emitting different events
> 3. We ensure we receive the expected event type from the expected handler filtering out those messages
> 
> Or did I miss it?

This is kinda done with my recent update of the predicate selection tests where 3 message handlers are configured, (1 being without the specific Service Bus extension) and 2 with different message types and predicate selection on a property on the message context.
